baybe/simulation/lookup.py

```diff
@@ -1,201 +1,301 @@
-from typing import TYPE_CHECKING, Literal
+from collections.abc import Callable, Collection
+from typing import Literal
 
 import numpy as np
 import pandas as pd
 
-from baybe.campaign import Campaign
+from baybe.simulation._imputation import _impute_lookup
+from baybe.targets.base import Target
 from baybe.targets.enum import TargetMode
-from baybe.utils.dataframe import add_fake_results
-
-if TYPE_CHECKING:
-    from baybe.targets import NumericalTarget
 
 _logger = logging.getLogger(__name__)
 
 
-def _look_up_target_values(
+def look_up_targets(
     queries: pd.DataFrame,
-    campaign: Campaign,
-    lookup: pd.DataFrame | Callable | None = None,
+    targets: Collection[Target],
+    lookup: pd.DataFrame | Callable | None,
     impute_mode: Literal[
         "error", "worst", "best", "mean", "random", "ignore"
     ] = "error",
-):
-    """Fill the target values in the query dataframe using the lookup mechanism.
+) -> None:
+    """Add/fill target values in a dataframe using a lookup mechanism.
 
-    Note that this does not create a new dataframe but modifies ``queries`` in-place.
+    Note:
+        This does not create a new dataframe but modifies ``queries`` in-place.
 
     Args:
-        queries: A dataframe containing points to be queried.
-        campaign: The campaign for which the experiments should be simulated.
-        lookup: The lookup mechanism.
-            See :func:`baybe.simulation.scenarios.simulate_scenarios` for details.
-        impute_mode: The used impute mode.
-            See :func:`baybe.simulation.scenarios.simulate_scenarios` for details.
+        queries: The dataframe to be modified. Its content must be compatible with the
+            chosen lookup mechanism.
+        targets: The targets whose values are to be looked up.
+        lookup: The lookup mechanism. Can be one of the following choices:
+
+            -   A dataframe mapping rows of ``queries`` to the corresponding target
+                values. That is, it must contain the same columns as ``queries`` plus
+                one additional column for each of the given target.
+            -   A callable, providing target values for each row of ``queries``.
+            -   ``None``. Produces fake values for all targets.
+        impute_mode: Specifies how a missing lookup will be handled. Only relevant for
+            dataframe lookups. Can be one of the following choices:
+
+            - ``"error"``: An error will be thrown.
+            - ``"worst"``: Imputes the worst available value for each target.
+            - ``"best"``: Imputes the best available value for each target.
+            - ``"mean"``: Imputes the mean value for each target.
+            - ``"random"``: A random row will be used for the lookup.
 
     Raises:
-        AssertionError: If an analytical function is used and an incorrect number of
-            targets was specified.
+        ValueError: If an invalid lookup mechanism is provided.
+
+    Example:
```

Comment:
tbh it can be very useful, nothing gets you up to speed than a usage example, imo its even better than the args and return texts, we simply dont have it a lot because its lots of work :D

Reply:
Would say the same: having examples is really great but tons of work. For some reason, I was in the mood to add it, and so I did. Would be nice to have more of them, especially for low level routines. But probably one should dive a bit deeper in the doctest plugin for that at some point and see how to cleverly organize imports etc. in order to keep the examples truly minimal

And: yes, the code is tested automatically.